### ISSUE TITLE
Functions with std::string/std::list parameters are never instrumented by LLVM backend

### ISSUE THREAD

**Issue by Teemperor** [AFLplusplus/AFLplusplus#1927]
**IMPORTANT**
1. You have verified that the issue to be present in the current `dev` branch.
  * ye
2. Please supply the command line options and relevant environment variables,
   e.g., a copy-paste of the contents of `out/default/fuzzer_setup`..
  * not even starting the fuzzer, N/A

**Describe the bug**
If a function contains a std::string/std::list parameter (or template class referencing those types), the function is not instrumented.

**To Reproduce**
1. Build AFL++. The system compiler should use libstdc++ by default (which is the case on most linux distros).
2. Create the following file:

```cpp
#include <cstdio>
#include <string>

#ifdef I_WANNA_DIE
typedef std::string Param;
#else
typedef int Param;
#endif

int g;

void func(Param p) {
  if (g) { // this probably should create coverage...
    printf("foo");
  }
}
```
3. Compile it with `./AFLplusplus/afl-clang-fast++ path/to/file.cpp` to see AFL++ would instrument this.
4. Run `./AFLplusplus/afl-clang-fast++ path/to/file.cpp -DI_WANNA_DIE=1` and it no longer instruments it.

```
$ ./AFLplusplus/afl-clang-fast++ what-the-fuck.cpp -o /dev/null -c
afl-cc++4.09a by Michal Zalewski, Laszlo Szekeres, Marc Heuse - mode: LLVM-PCGUARD
SanitizerCoveragePCGUARD++4.09a
[+] Instrumented 3 locations with no collisions (non-hardened mode) of which are 0 handled and 0 unhandled selects.
$ ./AFLplusplus/afl-clang-fast++ -DI_WANNA_DIE=1 what-the-fuck.cpp -o /dev/null -c
afl-cc++4.09a by Michal Zalewski, Laszlo Szekeres, Marc Heuse - mode: LLVM-PCGUARD
SanitizerCoveragePCGUARD++4.09a
[!] WARNING: No instrumentation targets found.
```

**Expected behavior**
It probably should instrument the function.

**Additional context**
The function `isIgnoreFunction` in  `instrumentation/afl-llvm-common.cc` contains the following code which seems to filter all these functions out because `std::string` is mangled as `St7__cxx1112basic_string...` with libstdc++ (the `__cxx` is the relevant substring).

```cpp
  static constexpr const char *ignoreSubstringList[] = {

      "__asan", "__msan",       "__ubsan",    "__lsan",  "__san", "__sanitize",
      "__cxx",  "DebugCounter", "DwarfDebug", "DebugLoc"

  };

  for (auto const &ignoreListFunc : ignoreSubstringList) {

    // hexcoder: F->getName().contains() not avaiilable in llvm 3.8.0
    if (StringRef::npos != F->getName().find(ignoreListFunc)) { return true; }

  }

```

TLDR: Filtering based on whether function name contains `__cxx` is skipping everything that referenced std::string/list in the mangled name.

**Comment by vanhauser-thc** [comment]
I have to filter out cxx stuff because they can be used in very early constructors which can result in a crash if I do not handle them.
It is rare that std:: variables are used in an if() and that the if is important (usually it is a return).
That is the reason I put it like that.

I will need to check if I can relax this.

**Comment by Teemperor** [comment]
I have the feeling that a string parameter in a function is not a rare corner case, but I'm no 10x developer.

FWIW, in a library such as [nlohmann::json](https://github.com/nlohmann/json), this bug removes 95% of all coverage (none of the skipped functions is from the standard library or a global initializer).

```bash
$ # Current version
$ ../old/afl-clang-fast++ fuzzer-parse_json.cpp -fsanitize=fuzzer
afl-cc++4.09a by Michal Zalewski, Laszlo Szekeres, Marc Heuse - mode: LLVM-PCGUARD
[+] Found '-fsanitize=fuzzer', replacing with libAFLDriver.a
SanitizerCoveragePCGUARD++4.09a
[+] Instrumented 234 locations with no collisions (non-hardened mode) of which are 11 handled and 0 unhandled selects.
$ # Without the `__cxx` check:
$ ../fixed/afl-clang-fast++ fuzzer-parse_json.cpp -fsanitize=fuzzer
afl-cc++4.09a by Michal Zalewski, Laszlo Szekeres, Marc Heuse - mode: LLVM-PCGUARD
[+] Found '-fsanitize=fuzzer', replacing with libAFLDriver.a
SanitizerCoveragePCGUARD++4.09a
[+] Instrumented 4605 locations with no collisions (non-hardened mode) of which are 199 handled and 0 unhandled selects.
```

**Comment by andreafioraldi** [comment]
i don't think marc had time to check this, this is not related to the prefix check but the contains one, thus is filtering out any mangled functions containing __cxx_, not branches using std:: vars (that are unlikely). This remained uncatched as we mostly use libc++ I guess on oss-fuzz et al., I'll reproduce and fix it asap.

**Comment by vanhauser-thc** [comment]
the fix is in, the coverage difference in general is only 0.15% though. let's see the reports where this breaks and then do a more fine-grained exception.
